server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -23,30 +23,39 @@
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
 /**
  * A logger that logs deprecation notices. Logger should be initialized with a parent logger which name will be used
  * for deprecation logger. For instance <code>new DeprecationLogger("org.elasticsearch.test.SomeClass")</code> will
  * result in a deprecation logger with name <code>org.elasticsearch.deprecation.test.SomeClass</code>. This allows to use
  * <code>deprecation</code> logger defined in log4j2.properties.
- *
+ * <p>
  * Deprecation logs are written to deprecation log file - defined in log4j2.properties, as well as warnings added to a response header.
  * All deprecation usages are throttled basing on a key. Key is a string provided in an argument and can be prefixed with
  * <code>X-Opaque-Id</code>. This allows to throttle deprecations per client usage.
  * <code>deprecationLogger.deprecate("key","message {}", "param");</code>
- *
- * @see ThrottlingAndHeaderWarningLogger for throttling and header warnings implementation details
+ * <p>
+ * Additional log handlers can be registered via {@link #addHandler(DeprecatedLogHandler)}. Subject to rate limiting,
+ * each additional handler will also be called with a deprecated message and its key.
  */
 public class DeprecationLogger {
-    private final ThrottlingAndHeaderWarningLogger deprecationLogger;
+    /** Log handlers are not tied to this specific <code>DeprecationLogger</code> instance */
+    private static final List<DeprecatedLogHandler> additionalHandlers = Collections.synchronizedList(new ArrayList<>());
```

Comment:
I can't reference `DeprecationIndexingService` directly because it only exists in X-Pack now. I don't expect multiple handlers, but I used a collection because it felt a bit wrong to have a mechanism for generic handlers but only support one.

Regarding synchronisation, I don't have strong opinions on the implementation since realistically, we'll either never modify the collection or only add a single handler.